gcc/cmassoc/mp.1
.TH mp 1 "August 2013" "cmassoc-tools-1.9.1" "Motley Tools"

.SH NAME
mp - format man page header

.SH SYNOPSIS
.BR mp
.RI [ options ]
.IR file
.RI [ file ] 
.RI [ ... ]

.PP
.BR mp
.RI [ options ]
.RI [< file ]
.RI [> file ]

.SH DESCRIPTION

.PP
Format Linux man pages to update header information, simplify text maintenance and correct common mistakes.

.PP
The man page filename and extension, taken from the command line, are used as the program name and folder number.
The title header, starting with "\fB.TH\fR", is updated using information from a user-defined configuration profile.
An empty line is inserted prior to each ".SH", ".TP" and ".PP" tag.
Sentences are each moved to a new line so that they can be easily edited or detected and replaced or removed by other programs.
Selected headers are detected and updated or preceded with one blank line.

.PP
This program can be used with program \fBnp\fR, for example, to perform global changes across multiple man pages.

.SH OPTIONS
Long options are not supported.
All options are single letters.

.TP
.RB - o
Print a template profile on stdout.
The template uses default program values.
Output can be directed to a file, edited and read back whenever the program is next executed.
This option takes precedence over normal formatting operations.

.TP
-\fBp\fI profile\fR
The profile name.
If this option is present then the program attempts to open and read the profile.
The program will report an error and terminated if the named profile file does not exist.
If this option is omitted the program attempts to read the default profile.
The program does not report an error when the default profile is missing.
The default profile is "\fB/etc/mp.ini\fR".

.TP
-\fBs\fI section\fR
The profile section name.
If this option is present the the program will search the named profile or default profile for the named section.
If this option is omitted the progam will search for the default section, instead.
The default section is named "\fBdefault\fR".

.TP
.RB - ? ,-- help
Print program help summary on stdout.
This option takes precedence over previous command line options.

.TP
.RB - ! ,-- version
Print program version information on stdout.
This option takes precedence over previous command line options.

.SH ARGUMENTS

.TP
.IR file
The name of man page.
This file is replaced and a numbered backup is created.
Use program \fBrex\fR to recover a prior version.

.SH PROFILE
The default profile looks like:

.PP
   [default]
   project=Motley Tools
   package=motley-tools-1.5.9
   release=Feb 2004

.TP
.BR project
The project name appears at the top center of the page.
A reasonable value is the company or project name.
The default is "Motley Tools".

.TP
.BR package
The package name appears in the lower left of the page.
A reasonable value is the archive name with revision numbers.
The default is "cmassoc-tools-x.x.x".

.TP
.BR release
The release data appears at the bottom center of the page.
A reasonable value is the current month and year.
The default is the current month and year.

.SH EXAMPLES
The following example formats the ".TH" line of man page \fBmp.7\fR using default program values.

.PP
   # mp afm.7

.SH SEE ALSO
.BR afm ( 1 ),
.BR cfm ( 1 ),
.BR np ( 1 ),
.BR rex ( 1 )

.SH CREDITS
 Charles Maier <cmaier@cmassoc.net>



gcc/cmassoc/mp.c
/*====================================================================*
 *
 *   mp.c - man page header;
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

#define _GETOPT_H

/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <time.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/console.h"
#include "../tools/profile.h"
#include "../tools/error.h"
#include "../tools/sizes.h"
#include "../tools/chars.h"
#include "../files/files.h"
#include "../tidy/tidy.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/console.c"
#include "../tools/profile.c"
#include "../tools/totruth.c"
#include "../tools/error.c"
#endif

#ifndef MAKEFILE
#include "../files/vfopen.c"
#include "../files/splitpath.c"
#include "../files/mergepath.c"
#include "../files/makepath.c"
#endif

#ifndef MAKEFILE
#include "../tidy/keep.c"
#include "../tidy/literal.c"
#include "../tidy/escaped.c"
#endif

/*====================================================================*
 *   program constants;
 *--------------------------------------------------------------------*/

#define MP_PROFILE "/etc/mp.ini"
#define MP_SECTION "default"
#define MP_PROGRAM ""
#define MP_PROJECT PROJECT
#define MP_PACKAGE PACKAGE "-" VERSION 
#define MP_PUBLISH "%b %Y"

#define MP_VERBOSE (1 << 0)
#define MP_SILENCE (1 << 1)

/*====================================================================*
 *
 *   void example ()
 *
 *   print a skeleton manpage example on stdout;
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void example () 

{ 
	char const * subjects [] = 
	{ 
		"NAME", 
		"SYNTAX", 
		"SYNOPSIS", 
		"AVAILABILITY", 
		"DESCRIPTION", 
		"OPTIONS", 
		"EXAMPLES", 
		"NOTES", 
		"MESSAGES", 
		"HISTORY", 
		"RESOURCES", 
		"FILES", 
		"BUGS", 
		"CAVEATS", 
		"AUTHORS", 
		"CREDITS", 
		"SEE ALSO", 
		(char const *)(0)
	}; 
	char const ** subject = subjects; 
	printf (".TH program 1 package"); 
	while (* subject) 
	{ 
		printf ("\n.SH %s\n", * subject++); 
	} 
	return; 
} 

/*====================================================================*
 *
 *   void configure (char const * section, char const * project, char const * package, char const * release) 
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 * 
 *--------------------------------------------------------------------*/

static void configure (char const * section, char const * project, char const * package, char const * release) 

{ 
	printf ("[%s]\n", section); 
	printf ("project=%s\n", project); 
	printf ("package=%s\n", package); 
	printf ("release=%s\n", release); 
	return; 
} 

/*====================================================================*
 *
 *   void function (char const * program, char const * project, char const * package, char const * release, flag_t flags);
 *
 *
 *   by convention, man pages are placed in /usr/share/man/man(n) 
 *   where (n) is one of the following topics. Consequently, name
 *   your manpage as <topic>.<n> where <n> is the folder number;
 *
 *   1 Executable Shell Commands
 *   2 System Calls
 *   3 Library Calls
 *   4 Devices, Network Interfaces
 *   5 File Formats, Conventions
 *   6 Games, Demonstration Software
 *   7 Miscellaneous
 *   8 Maintenance Commands (for root user only)
 *   9 Kernel/X Windows Information
 *
 *
 *   to format man pages use one of the following Linux commands.
 *
 *   nroff -man <manpage> > <text_file>
 *   groff -man -Tascii <manpage> <ascii_text_file>
 *   groff -man <manpage> <postscript_file>
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 * 
 *--------------------------------------------------------------------*/

static void function (char const * program, char const * project, char const * package, char const * release, flag_t flags) 

{ 
	signed c = getc (stdin); 
	while (c != EOF) 
	{ 
		while ((c == '\r') || (c == '\n')) 
		{ 
			c = getc (stdin); 
		} 
		if (c == '.') 
		{ 
			char symbol [SYMBOLSIZE]; 
			char * sp = symbol; 
			do 
			{ 
				* sp++ = c; 
				c = getc (stdin); 
			} 
			while (isalpha (c)); 
			* sp = (char) (0); 
			if (!strcmp (symbol, ".TH")) 
			{ 
				while (nobreak (c)) 
				{ 
					c = getc (stdin); 
				} 
				sp += snprintf (sp, sizeof (symbol) + sp - symbol, " %s", program); 
				sp += snprintf (sp, sizeof (symbol) + sp - symbol, " \"%s\"", release); 
				sp += snprintf (sp, sizeof (symbol) + sp - symbol, " \"%s\"", package); 
				sp += snprintf (sp, sizeof (symbol) + sp - symbol, " \"%s\"", project); 
			} 
			else if (!strcmp (symbol, ".SH")) 
			{ 
				putc ('\n', stdout); 
			} 
			else if (!strcmp (symbol, ".SS")) 
			{ 
				putc ('\n', stdout); 
			} 
			else if (!strcmp (symbol, ".TP")) 
			{ 
				putc ('\n', stdout); 
			} 
			else if (!strcmp (symbol, ".HP")) 
			{ 
				putc ('\n', stdout); 
			} 
			else if (!strcmp (symbol, ".IP")) 
			{ 
				putc ('\n', stdout); 
			} 
			else if (!strcmp (symbol, ".PP")) 
			{ 
				putc ('\n', stdout); 
			} 
			else if (!strcmp (symbol, ".P")) 
			{ 
				putc ('\n', stdout); 
			} 
			fputs (symbol, stdout); 
		} 
		while (nobreak (c)) 
		{ 
			if (isquote (c)) 
			{ 
				c = literal (c); 
				continue; 
			} 
			if (c == '.') 
			{ 
				c = keep (c); 
				if (c == '.') 
				{ 
					do 
					{ 
						c = keep (c); 
					} 
					while (c == '.'); 
					continue; 
				} 
				if (isblank (c)) 
				{ 
					do 
					{ 
						c = getc (stdin); 
					} 
					while (isblank (c)); 
					putc ('\n', stdout); 
					continue; 
				} 
				continue; 
			} 

#if 1

			if (c == '\\') 
			{ 
				c = keep (c); 
				if (c == 'v') 
				{ 
					c = 'f'; 
				} 
				c = keep (c); 
				continue; 

#endif

			} 
			c = keep (c); 
		} 
		c = keep (c); 
	} 
	putc ('\n', stdout); 
	return; 
} 

/*====================================================================*
 *
 *   int main (int argc, char const * argv []);
 *
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv []) 

{ 
	static char const * optv [] = 
	{ 
		"mop:s:", 
		"tidy Linux manpage", 
		PUTOPTV_S_FILTER, 
		"m\tprint example manpage on stdout", 
		"o\tprint default profile on stdout", 
		"p f\tprofile is (s) [" LITERAL (MP_PROFILE) "]", 
		"s s\tsection is (s) [" LITERAL (MP_SECTION) "]", 
		(char const *) (0)
	}; 
	char buffer [16]; 
	time_t now = time ((time_t *)(0)); ; 
	char const * profile = MP_PROFILE; 
	char const * section = MP_SECTION; 
	char const * program = MP_PROGRAM; 
	char const * project = MP_PROJECT; 
	char const * package = MP_PACKAGE; 
	char const * release = buffer; 
	char * sp; 
	flag_t flags = (flag_t)(0); 
	signed c; 
	strftime (buffer, sizeof (buffer), MP_PUBLISH, localtime (& now)); 
	while (~ (c = getoptv (argc, argv, optv))) 
	{ 
		switch (c) 
		{ 
		case 'p': 
			profile = optarg; 
			break; 
		case 's': 
			section = optarg; 
			break; 
		case 'm': 
			example (release); 
			return (0); 
		case 'o': 
			configure (section, project, package, release); 
			return (0); 
		default: 
			break; 
		} 
	} 
	argc -= optind; 
	argv += optind; 
	project = profilestring (profile, section, "project", project); 
	package = profilestring (profile, section, "package", package); 
	release = profilestring (profile, section, "release", release); 
	if ((!argc) || (!* argv)) 
	{ 
		function ("unamed 7", project, package, release, flags); 
	} 
	while ((argc) && (* argv)) 
	{ 
		if (vfopen (* argv)) 
		{ 
			program = * argv; 
			for (sp = (char *)(* argv); * sp; ++ sp) 
			{ 
				if (* sp == PATH_C_EXTENDER) 
				{ 
					program = sp + 1; 
				} 
				if (* sp == FILE_C_EXTENDER) 
				{ 
					* sp = ' '; 
				} 
			} 
			function (program, project, package, release, flags); 
		} 
		argc--; 
		argv++; 
	} 
	exit (0); 
} 

